### PR TITLE
Fix path params being in the body

### LINKED ISSUE #201: Fields bound by path template also in body (Swagger)
https://github.com/googleapis/googleapis/blob/master/google/api/http.proto#L123

| HTTP | RPC |
| --- | --- |
| `PUT /v1/messages/123456 { "text": "Hi!" }` | `UpdateMessage(message_id: "123456" message { text: "Hi!" })` |

The special name `*` can be used in the body mapping to define that
every field not bound by the path template should be mapped to the
request body.  This enables the following alternative definition of
the update method:

``` proto
service Messaging {
  rpc UpdateMessage(Message) returns (Message) {
    option (google.api.http) = {
      put: "/v1/messages/{message_id}"
      body: "*"
    };
  }
}
message Message {
  string message_id = 1;
  string text = 2;
}
```

---

Currently, this doesn't seem to work when generating the Swagger definition, as both fields are showing up. This is a snippet from my `.proto` file.

``` proto
service Application {
    rpc Update(UpdateApplicationRequest) returns (UpdateApplicationResponse) {
        option (google.api.http) = {
            put: "/api/v1/application/{appEUI}"
            body: "*"
        };
    }
}

message UpdateApplicationRequest {
    string appEUI = 1;
    string name = 2;
}

```

image

I would have expected to ony have `name` in the request body. Leaving out the `devEUI` does work however when making requests!

When setting `body: "name"` in the `.proto` file (see also [http.proto#L109](https://github.com/googleapis/googleapis/blob/master/google/api/http.proto#L109)), the API returns the following error (even if the `appEUI` is included in the body too):

``` json
{
  "Error": "json: cannot unmarshal object into Go value of type string",
  "Code": 3
}
```

### PR DESCRIPTION
**Pull request by stijndehaes**
#### References to other Issues or PRs

Fixes #201

#### Have you read the [Contributing Guidelines](https://github.com/grpc-ecosystem/grpc-gateway/blob/master/CONTRIBUTING.md)?

I have read it, and have regenerated the files with the first and second docker run command. 

#### Brief description of what is fixed or changed

I toke the previous PR here: https://github.com/grpc-ecosystem/grpc-gateway/pull/1145

And migrated it to the current version of the code. I also tested this out on a project of mine, and the result looks good.

#### Other comments

~I am not sure how to unit test this. If somebody could give me some hints that would be wonderful.~
I added one unit test, not sure if I added it to the right place though. I would like feedback on this :)

### CODE REVIEW

**Review by johanbrandhorst — COMMENTED** [review]
Thanks for picking this up @stijndehaes. @jgiles, since this is based on your original contribution, could you please explicitly give consent to have it be merged here?

File: protoc-gen-openapiv2/internal/genopenapi/template.go [hunk 1]
```diff
@@ -384,7 +375,92 @@ func skipRenderingRef(refName string) bool {
...
+		}
+
+		kv := keyVal{Value: fieldValue}
+		kv.Key = reg.FieldName(f)
+		if schema.Properties == nil {
+			schema.Properties = &openapiSchemaObjectProperties{}
+		}
+		*schema.Properties = append(*schema.Properties, kv)
+	}
+	return schema
```
[the remark this answers is not in the document]
Comment (stijndehaes) [reply protoc-gen-openapiv2/internal/genopenapi/template.go:460]: I quickly tried this. But for example in the `flow_combination.proto` there are empty responses, by making this method return an error that is a problem. I added the check where the method is used to generate the body instead. I hope this makes sense, if you would like to do this another way I can change it

File: protoc-gen-openapiv2/internal/genopenapi/template.go [hunk 2]
```diff
@@ -986,14 +986,17 @@ func renderServices(services []*descriptor.Service, paths openapiPathsObject, re
 								}
 							}
 							if len(bodyExcludedFields) != 0 {
-								schema = renderMessageAsDefinition(meth.RequestType, reg, customRefs, bodyExcludedFields)
+								msgDef := renderMessageAsDefinition(meth.RequestType, reg, customRefs, bodyExcludedFields)
+								if msgDef.Properties == nil || len(*msgDef.Properties) == 0 {
+									return fmt.Errorf("not possible to create a schema for object with 0 properties for message: %s", *meth.RequestType)
+								}
```
Comment (johanbrandhorst) [inline protoc-gen-openapiv2/internal/genopenapi/template.go:992]: Thanks for this change, though I just had another thought about this. I would see this as a bug fix, but it's also probably a breaking change for users, so maybe we should just log a warning instead. See `glog.Errorf` uses in this file.

File: examples/internal/proto/examplepb/a_bit_of_everything.proto [hunk 3]
```diff
@@ -641,8 +641,7 @@ service ABitOfEverythingService {
 	}
 	rpc PostWithEmptyBody(Body) returns (google.protobuf.Empty) {
 		option (google.api.http) = {
-			post: "/v2/example/postwithemptybody/{name}",
-			body: "*"
```
Comment (johanbrandhorst) [inline examples/internal/proto/examplepb/a_bit_of_everything.proto:645]: Lets revert this change too, apparently this is behaviour we want to keep working, judging by https://github.com/grpc-ecosystem/grpc-gateway/pull/527. Sorry for the confusion.

**Review by johanbrandhorst — COMMENTED** [review]
I changed my mind, it would be a breaking change to make this error, so lets just leave it as it was, sorry for the confusion.

File: examples/internal/proto/examplepb/a_bit_of_everything.swagger.json [hunk 4]
```diff
@@ -1906,7 +1906,202 @@
             "in": "body",
             "required": true,
             "schema": {
-              "$ref": "#/definitions/examplepbABitOfEverything"
+              "type": "object",
+              "example": {
+                "uuid": "0cf361e1-4b44-483d-a159-54dabdf7e814"
+              },
```
[the remark this answers is not in the document]
Comment (stijndehaes) [reply examples/internal/proto/examplepb/a_bit_of_everything.swagger.json]: I changed the example to include `int64_value` and `double_value`

File: examples/internal/proto/examplepb/a_bit_of_everything.proto (hunk 3 above)
Comment (stijndehaes) [reply examples/internal/proto/examplepb/a_bit_of_everything.proto:645]: No problem, I'll change it

**Review by johanbrandhorst — APPROVED** [review]
LGTM, however I'd like to wait for @jgiles to have a chance to give consent to this being merged since it contains their original work.

**Comment by stijndehaes** [comment]
@jgiles A friendly reminder to give your consent to merge this as I based this work on your original PR :)

**Comment by jgiles** [comment]
I consent to merging the code. Thank you for picking this up, and apologies for abandoning the original PR :-)

**Comment by jgiles** [comment]
@stijndehaes thank you!

**Comment by johanbrandhorst** [comment]
Thanks for your contributions, @stijndehaes and @jgiles :)!

**Comment by stijndehaes** [comment]
@jgiles without your original code I could have never landed this so easily so thank you for starting and giving your consent ;)

**Pull request merged**